core/src/actors/codec.rs

```diff
@@ -1,49 +1,52 @@
 use std::io;
+use std::io::Cursor;
+
+use byteorder::{BigEndian, ReadBytesExt, WriteBytesExt};
 
 use actix::Message;
 use bytes::BytesMut;
 use log::info;
 use tokio::codec::{Decoder, Encoder};
 
+const HEADER_SIZE: u16 = 2; // bytes
+
 /// Message coming from the network
-#[derive(Debug, Message)]
+#[derive(Debug, Message, Eq, PartialEq, Clone)]
 pub enum Request {
     /// Request message
-    Message(String),
+    Message(BytesMut),
 }
 
 /// Message going to the network
-#[derive(Debug, Message)]
+#[derive(Debug, Message, Eq, PartialEq, Clone)]
 pub enum Response {
     /// Response message
-    Message(String),
+    Message(BytesMut),
 }
 
 /// Codec for client -> server transport
+#[derive(Debug, Message, Eq, PartialEq, Clone)]
 pub struct P2PCodec;
 
 /// Implement decoder trait for P2P codec
+///
 impl Decoder for P2PCodec {
     type Item = Request;
     type Error = io::Error;
 
     /// Method to decode bytes to a request
     fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
-        // Locate a byte corresponding to a '\n' in the byte stream
-        if let Some(i) = src.iter().position(|&b| b == b'\n') {
-            // Remove the serialized frame from the buffer.
-            let line = src.split_to(i + 1);
-
-            // Parse the buffer as an UTF-8 encoded string
-            let mut res = String::from_utf8(line.to_vec()).unwrap();
-
-            // Remove the last two bytes of the string (corresponding to \r\n)
-            res.truncate(res.len() - 2);
-
-            Ok(Some(Request::Message(res)))
-        } else {
-            Ok(None)
+        let mut ftb: Option<Self::Item> = None;
+        let msg_len = src.len() as u16;
+        if msg_len > 2 {
+            let mut header_vec = Cursor::new([src.to_vec()[0], src.to_vec()[1]]);
+            let msg_size = header_vec.read_u16::<BigEndian>().unwrap();
+            if msg_len >= (msg_size + HEADER_SIZE) {
+                src.split_to(HEADER_SIZE as usize);
```

Comment:
is it proper to add comments here indicating how the byte stream is manipulated?:
```
// trim the header from given byte stream
// [000 00100] => [00100]
//  ^^^
//  |-- header
```